### ISSUE TITLE
Dispatch Thread Dead Unexpectedly

### ISSUE THREAD

**Issue by ylgrgyq** [relayrides/pushy#144]
![2014-12-08 10 26 47](https://cloud.githubusercontent.com/assets/1115061/5334130/ddcc22cc-7ec5-11e4-93c6-4982dcca00fc.png)
I never reproduce this problem in sand box environment. The sendAttemptLimit is set to 8000. So every 8000 msgs sent to APNS successfully, ApnsConnection will be reconnected. 
The stack trace is thrown by shutdownGracefully() method in ApnsConnection class. It's code is as follows:
![image](https://cloud.githubusercontent.com/assets/1115061/5338114/7236fdfc-7f0b-11e4-879b-db22eb7bb55c.png)

**Comment by killme2008** [comment]
I have the same exception in our push service:

```
Dispatch thread for 0nvnphkmxw6m17ujypcajmg2iib5nn9fp1rjogd26wy93y09:gateway.push.apple.com died unexpectedly. Please file a bug with the exception details.
java.util.NoSuchElementException: idleStateHandler
	at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:907)
	at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:304)
	at com.relayrides.pushy.apns.ApnsConnection.shutdownGracefully(ApnsConnection.java:626)
	at com.relayrides.pushy.apns.ApnsConnection.sendNotification(ApnsConnection.java:581)
	at com.relayrides.pushy.apns.PushManager$1.run(PushManager.java:277)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by killme2008** [comment]
Any progress for this issue?
We will add a `try catch` block for removing handler to fix it temporary.

**Comment by jchambers** [comment]
Can you tell me more about your configuration? Have you configured connections to close after a period of inactivity? Does this happen every time a connection closes?

Thanks!

**Comment by killme2008** [comment]
The configuration as below(in clojure):

```clojure
(doto
      (PushManagerConfiguration.)
    (.setConcurrentConnectionCount 2)
    (.setConnectionConfiguration
     (doto (ApnsConnectionConfiguration.)
       (.setSendAttemptLimit (Integer/valueOf 10)) 
       (.setCloseAfterInactivityTime (Integer/valueOf 300))
       (.setGracefulShutdownTimeout (Integer/valueOf 100))))))
```

We configure connections to be closed after 300 seconds of inactivity.It doesn't happen every time when a connection closes,but it happens every time when we want to push too many notifications (over 1 millions)

**Comment by jchambers** [comment]
@killme2008 As a bit of a science experiment, would you mind giving things a try without direct buffers? @sunng87 [explains how to disable direct buffers](https://github.com/relayrides/pushy/issues/142#issuecomment-65749004) in a separate issue.

Thanks!

**Comment by killme2008** [comment]
Hi, @jchambers 
We disables direct buffer right now `-Dio.netty.noPreferDirect=true`

P.S.  sunny is my colleague.

**Comment by jchambers** [comment]
It seems like there's a chance this might have been fixed by https://github.com/netty/netty/issues/3364. We bumped to Netty 4.0.26 in #154; would you mind giving that a shot and seeing if it resolves this issue?

Thanks for your patience!

**Comment by killme2008** [comment]
OK, thanks for your reminding.We will  try it.

2015-03-04 1:39 GMT+08:00 Jon Chambers <notifications@github.com>:

> It seems like there's a chance this might have been fixed by
> netty/netty#3364 <https://github.com/netty/netty/issues/3364>. We bumped
> to Netty 4.0.26 in #154 <https://github.com/relayrides/pushy/pull/154>;
> would you mind giving that a shot and seeing if it resolves this issue?
>
> Thanks for your patience!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/relayrides/pushy/issues/144#issuecomment-76994523>.
>



-- 
庄晓丹
Email:        killme2008@gmail.com xzhuang@avos.com
Site:           http://fnil.net
Twitter:      @killme2008

**Comment by jchambers** [comment]
Looks like this was just a weird race condition. Although I haven't been able to reproduce the failure you're seeing, it *should* be fixed by ffc6fea.
